File: contrib/bin/xgbd/xgbd_evaluate.rb
#!/usr/bin/env ruby

# Evaluate an xgboost model

require 'set'
require 'google/protobuf'

c3tk_home = ENV['C3TK_HOME']
raise 'C3TK_HOME not defined' unless c3tk_home

require "#{c3tk_home}/bin/ruby/lib/iwcmdline"
require "#{c3tk_home}/bin/py/pytk/xgbd/xgboost_model_pb"

def usage
msg = <<-END
Scores an xgboost descriptor model built with xgbd_make.
Takes either a smiles file or a descriptor file,
If a smiles file is given, the descriptors required for the model will be computed by make_descriptors
  -mdir <dir>           model directory created by xgbd_make
  -smi                  input is a smiles file - model descriptors will be computed.
  -pipe                 use descriptor_pipeline for descriptor compuation,
                        only some descriptors are supported, but maybe faster.
  -j <threads>          send the -j option to make_descriptors.sh
  -v                    verbose output
END
  $stderr << msg

  exit(0)
end

# Return the proto mdir/model_metadata.dat
def get_model_metadata(mdir)
  fname = File.join(mdir, 'model_metadata.dat')

  unless File.size?(fname)
    $stderr << "Empty or missing model metadata #{fname}\n"
    return 1
  end

  return XgboostModel::XGBoostModel.decode(File.read(fname))
end

def xgbd_evaluate_smiles(fname, mdir, proto, cl)
  descriptors = Set.new()
  proto.name_to_col.each do |name, col|
    descriptors.add(name.gsub(/_.*/, ""))
  end

  if cl.option_present('pipe')
    cmd = 'descriptor_pipeline.sh'
  else
    cmd = 'make_descriptors.sh'
  end

  if cl.option_present('j')
    j = cl.value('j')
    cmd << " -j #{j}"
  end

  descriptors.each do |d|
    cmd << " -#{d}"
  end
  cmd << " #{fname}"
  cmd << "| xgboost_model_evaluate.sh -mdir #{mdir} -"

  $stderr << "Executing #{cmd}\n" if cl.option_present('v')

  system(cmd)
end

def xgbd_evaluate_descriptors(fname, mdir, proto, cl)
  cmd = "xgboost_model_evaluate.sh -mdir #{mdir} #{fname}"

  $stderr << "Executing #{cmd}\n" if cl.option_present('v')
  system(cmd)
end

def main
  cl = IWCmdline.new("-v-mdir=dir-smi-pipe-j=ipos")

  if cl.unrecognised_options_encountered
    $stderr << "unrecognised_options_encountered\n"
    usage
  end

  unless cl.option_present('mdir')
    $stderr << "Must specify xgboost model directory via the -mdir option\n"
    usage
  end

  if ARGV.empty?
    $stderr << "Insufficient arguments\n"
    usage
  end

  mdir = cl.value('mdir')

  proto = get_model_metadata(mdir)

  if cl.option_present('smi') || ARGV[0].match(/\.smi$/)
    xgbd_evaluate_smiles(ARGV[0], mdir, proto, cl)
  else
    xgbd_evaluate_descriptors(ARGV[0], mdir, proto, cl)
  end

end

main

File: contrib/bin/xgbd/xgbd_evaluate.py
# Evaluate an xgboost descriptor model built with xgboost_make

import os
import re

import pandas as pd

from absl import app
from absl import flags
from absl import logging
from google.protobuf import text_format

from xgboost import XGBClassifier
from xgboost import XGBRegressor

import xgboost_model_pb2

FLAGS = flags.FLAGS

flags.DEFINE_string("mdir", "", "Model directory")

def get_model(mdir: str)->tuple:
  """Look for 'what_kind_of_model` in `mdir` and make sure it is OK
    Return a model instantiated from mdir/xgboost.json and the
    name of the response
  """
  fname = os.path.join(mdir, "model_metadata.txt")
  if not os.path.exists(fname):
    logging.error("%s not found", fname)
    return None, None

  with open(fname, "r") as reader:
    text = reader.read()

  proto = text_format.Parse(text, xgboost_model_pb2.XGBoostModel())
  if not proto:
    logging.error("Cannot interpret as proto %s", text)
    return None, None

  if not proto.response:
    logging.error("No response in %s", fname)
    return None, None

  model_file = os.path.join(mdir, "xgboost.json")
  if not os.path.exists(model_file):
    logging.error("%s not found", model_file)
    return None

  model = XGBRegressor()
  model.load_model(model_file)

  return model, proto.response

def xgboost_evaluate(mdir: str, fname: str)->bool:
  """Read `fname` as descriptors for a model in `mdir`
  """
  if not os.path.isdir(mdir):
    logging.error("Model directory %s not found", mdir)
    return False

  model, response = get_model(mdir)
  if not model:
    logging.error("Invalid mode in %s", mdir)
    return False

  data = pd.read_csv(fname, sep=' ', header=0)

  logging.info("Evaluating %d rows", len(data))
  results = model.predict(data.iloc[:,1:])
  print(f"Id XGBD_{response}")
  for i in range(len(results)):
    print(f"{data.iloc[i,0]} {results[i]:.4f}")

  return True

def main(argv):
  """Evaluate an xgboost descriptor model.
  """
  if len(argv) == 1:
    logging.error("Must specify descriptor file as argument")
    return 1

  if not FLAGS.mdir:
    logging.error("must specify model directory via the --mdir option")
    return 1


  return xgboost_evaluate(FLAGS.mdir, argv[1])

if __name__ == '__main__':
  absl::InitializeLog()
  app.run(main)
